python/rainbow/simulators/proximal_contact/solver_interface.py

```diff
@@ -0,0 +1,133 @@
+import numpy as np
+from abc import ABC, abstractmethod
+from rainbow.util.timer import Timer
+
+
+class SolverInterface(ABC):
+    def __init__(self, J, WJT, b, mu, friction_solver, engine, stats, debug_on, prefix) -> None:
+        self.J = J # The contact jacobi matrix
+        self.WJT = WJT # The WJ^T matrix, here W = M^{-1}.
+        self.b = b # b = Ju^t + ∆t JM^{-1}h + EJu^t
+        self.mu = mu # The coefficient of friction.
+        self.friction_solver = friction_solver # The proximal operator of the friction cone.
+        self.engine = engine # The engine object.
+        self.stats = stats # The statistics information.
+        self.debug_on = debug_on # Whether to debug (Ture or False).
+        self.prefix = prefix # The prefix of the statistics information.
+        self.timer = None # The timer used to record the time
+        self.K = len(engine.contact_points) # Number of contact points.
+        self.iteration = 0 # The current iteration.
+        self.x = np.zeros(self.b.shape, dtype=np.float64)  # The current iterate
+        self.sol = np.zeros(
+            self.b.shape, dtype=np.float64
+        )  # The last best known solution, used for restarting if divergence
+        self.error = np.zeros(self.b.shape, dtype=np.float64)  # The residual vector
+        self.last_merit = np.Inf  # The last merit value
+        
+    def compute_initial_r_factor(self):
+        """ Compute the initial r-factor value.
+        """
+        delassus_diag = np.sum(self.J.multiply(self.WJT.T), axis=1).A1
+        delassus_diag[delassus_diag == 0] = 1
+        self.r = 0.1 / delassus_diag
+    
+    def update_r_factor(self):
```

Comment:
The function appears to do more than just update the r-factor

It appears to also update the solution vector with the current iterative if a non-divergent test is encountered.

Naming should be changed to reflect behavior of function

Reply:
Yeah, I will change the function name.